### PR TITLE
Websites: Add tunneling functionality

### PR DESCRIPTION
**Pull request by LukaszStem**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `./scripts/ci/test_static.sh` locally? (`pip install pylint flake8` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli-extensions/blob/master/docs/extension_summary_guidelines.md).

### CODE REVIEW

File: src/webapp/azext_webapp/tunnel.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+from websocket import create_connection, WebSocket
+
+from knack.util import CLIError
+from knack.log import get_logger
+logger = get_logger(__name__)  # TODO: Replace print with logger below
+
+class TunnelWebSocket(WebSocket):
+    def recv_frame(self):
+        frame = super(TunnelWebSocket, self).recv_frame()
+        print('Received frame:{}'.format(frame))
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: either remove this or have a --verbose flag that will print this.

File: src/webapp/azext_webapp/tunnel.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        return frame
+    
+    def recv(self):
+        data = super(TunnelWebSocket, self).recv()
+        print('Received websocket data:{}'.format(data))
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: either remove this or have a --verbose flag that will print this.

File: src/webapp/azext_webapp/tunnel.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        self.remote_addr = remote_addr
+        self.remote_user_name = remote_user_name
+        self.remote_password = remote_password
+        print('Creating a socket')
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        print('Setting socket options')
+        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        print('Binding to socket on local address and port')
+        self.sock.bind((self.local_addr, self.local_port))
+        print('Got to the end of init')
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: either remove this or have a --verbose flag that will print this.

File: src/webapp/azext_webapp/tunnel.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                                        enable_multithread=True)
+            print('websocket, connected status:{}', self.ws.connected)
+
+            index = index + 1
+            print('got debugger connection... index:{}'.format(index))
+            debugger_thread = Thread(target = self.listen_to_client, args = (self.client, self.ws, index))
+            web_socket_thread = Thread(target = self.listen_to_web_socket, args = (self.client, self.ws, index))
+            debugger_thread.start()
+            web_socket_thread.start()
+            print('both threads started...')
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: This print statement should be more user friendly.

File: src/webapp/azext_webapp/tunnel.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            debugger_thread.join()
+            web_socket_thread.join()
+            print('both threads stopped...')
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: This print statement should be more user friendly.

File: src/webapp/azext_webapp/tunnel.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    def listen(self):
+        self.sock.listen(100)
+        index = 0
+        basic_auth_string = self.create_basic_auth()
+        while True:
+            self.client, address = self.sock.accept()
+            self.client.settimeout(60)
+            host = 'wss://{}{}'.format(self.remote_addr,'.scm.azurewebsites.net/AppServiceTunnel/Tunnel.ashx')
+            basic_auth_header = 'Authorization: Basic {}'.format(basic_auth_string)
+            websocket.enableTrace(True)
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: either remove this or have a --verbose flag that will print this.

File: src/webapp/azext_webapp/tunnel.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            web_socket_thread.join()
+            print('both threads stopped...')
+
+    def listen_to_web_socket(self, client, ws_socket, index):
+        size = 4096
+        while True:
+            try:
+                print('Waiting for websocket data, connection status:{}, index:{}'.format(ws_socket.connected, index))
+                data = ws_socket.recv()
+                print('Received websocket data:{}, index:{}'.format(data, index))
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: either remove this or have a --verbose flag that will print this.

File: src/webapp/azext_webapp/tunnel.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                if data:
+                    # Set the response to echo back the recieved data 
+                    response = data
+                    print('Sending to debugger, response:{}, index{}'.format(response, index))
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: either remove this or have a --verbose flag that will print this.

File: src/webapp/azext_webapp/tunnel.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                    client.sendall(response)
+                    print('Done sending to debugger, index:{}', index)
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: either remove this or have a --verbose flag that will print this.

File: src/webapp/azext_webapp/tunnel.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+                return False
+
+    def listen_to_client(self, client, ws_socket, index):
+        size = 4096
+        while True:
+            try:
+                print('Waiting for debugger data, index:{}'.format(index))
+                buf = bytearray(4096)
+                nbytes = client.recv_into(buf, 4096)
+                print('Received debugger data, nbytes:{}, index:{}'.format(nbytes, index))
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: either remove this or have a --verbose flag that will print this.

File: src/webapp/azext_webapp/tunnel.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+                if nbytes > 0:
+                    responseData = buf[0:nbytes]
+                    print('Sending to websocket, response data:{}, index:{}'.format(responseData, index))
```
Comment (rramachand21) [inline src/webapp/azext_webapp/tunnel.py]: either remove this or have a --verbose flag that will print this.

File: src/webapp/azext_webapp/_help.py [hunk 12]
```diff
@@ -16,6 +16,15 @@
             az webapp up -n MyUniqueAppName --dryrun \n
             az webapp up -n MyUniqueAppName -l locationName
 """
+helps['webapp remote-connection'] = """
+    type: group
+    short-summary: Create a remote connection using a tcp tunnel to your app
```
Comment (panchagnula) [inline src/webapp/azext_webapp/_help.py]: both have the same help message. How is this different from create?

Comment (panchagnula) [reply src/webapp/azext_webapp/_help.py]: Will this work for function app as well? if not clarify in the help & i also recommend handling this (no support for function app) in the command with a useful error message

Comment (LukaszStem) [reply src/webapp/azext_webapp/_help.py]: The first help definition corresponds to the help for the group, while the 'create' command help is the help only for the command itself. It just so happened to be the same for both. Should I change this?

File: src/webapp/azext_webapp/custom.py [hunk 13]
```diff
@@ -213,3 +215,35 @@ def restore_webapp_snapshot(cmd, resource_group, name, time, slot=None, restore_
...
+    t.daemon = True
+    t.start()
+    if not _check_for_ready_tunnel(cmd, resource_group_name, name, config.remote_debugging_enabled, tunnel_server, slot):
+        print('Tunnel is not ready yet, please wait (may take up to 1 minute)')
+        while True:
+            time.sleep(1)
+            print('.')
+            if _check_for_ready_tunnel(cmd, resource_group_name, name, config.remote_debugging_enabled, slot):
+                break
+    print('Tunnel is ready! Creating on port {}'.format(port))
```
Comment (derekbekoe) [inline src/webapp/azext_webapp/custom.py]: Should use logger.warning for such messages. Same as above.

File: src/webapp/setup.py [hunk 14]
```diff
@@ -32,7 +32,7 @@
     description='An Azure CLI Extension to manage appservice resources',
     long_description='An Azure CLI Extension to manage appservice resources',
     license='MIT',
-    author='Sisira Panchagnula',
+    author='Sisira Panchagnula, Lukasz Stempniewicz',
```
Comment (panchagnula) [inline src/webapp/setup.py]: this is causing CI build break. Update the contacts under index.json as well.

**Pull request merged**